File: src/lib/api/tags.ts
import { invoke } from "@tauri-apps/api/core";

import type { RowSelection } from "./types";

export interface TagSummary {
  name: string;
  rowCount: number;
}

export interface TagSelectionSummary {
  name: string;
  selectedRows: number;
}

export interface TagMutationResult {
  affectedRows: number;
  normalizedTags: string[];
  associationsChanged: number;
}

export function listTags(): Promise<TagSummary[]> {
  return invoke<TagSummary[]>("list_tags");
}

export function createTag(name: string): Promise<boolean> {
  return invoke<boolean>("create_tag", { name });
}

export function deleteTag(name: string): Promise<boolean> {
  return invoke<boolean>("delete_tag", { name });
}

export function renameTag(oldName: string, newName: string): Promise<boolean> {
  return invoke<boolean>("rename_tag", { oldName, newName });
}

/** 最近使用的 Tag（打标对话框置顶用），持久化在 settings 表 */
export async function getRecentTags(): Promise<string[]> {
  const raw = await invoke<string>("get_recent_tags");
  if (!raw) {
    return [];
  }
  try {
    const parsed: unknown = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed.filter(item => typeof item === "string") : [];
  } catch {
    return [];
  }
}

export function setRecentTags(tags: string[]): Promise<void> {
  return invoke<void>("set_recent_tags", { json: JSON.stringify(tags) });
}

export function listSelectionTags(selection: RowSelection): Promise<TagSelectionSummary[]> {
  return invoke<TagSelectionSummary[]>("list_selection_tags", { selection });
}

export function addTagsToSelection(
  selection: RowSelection,
  tags: string[],
): Promise<TagMutationResult> {
  return invoke<TagMutationResult>("add_tags_to_selection", { selection, tags });
}

export function removeTagsFromSelection(
  selection: RowSelection,
  tags: string[],
): Promise<TagMutationResult> {
  return invoke<TagMutationResult>("remove_tags_from_selection", { selection, tags });
}

export function setTagsForRow(rowId: number, tags: string[]): Promise<TagMutationResult> {
  return invoke<TagMutationResult>("set_tags_for_row", { rowId, tags });
}

File: src-tauri/src/app/commands/tags.rs
use super::error_text;
use crate::app::AppRuntime;
use crate::db::{RowSelection, TagMutationResult, TagSelectionSummary, TagSummary};
use tauri::State;

#[tauri::command]
pub(crate) fn list_tags(runtime: State<'_, AppRuntime>) -> Result<Vec<TagSummary>, String> {
    runtime.list_tags().map_err(error_text)
}

#[tauri::command]
pub(crate) fn create_tag(name: String, runtime: State<'_, AppRuntime>) -> Result<bool, String> {
    runtime.create_tag(&name).map_err(error_text)
}

#[tauri::command]
pub(crate) fn delete_tag(name: String, runtime: State<'_, AppRuntime>) -> Result<bool, String> {
    runtime.delete_tag(&name).map_err(error_text)
}

#[tauri::command]
pub(crate) fn rename_tag(
    old_name: String,
    new_name: String,
    runtime: State<'_, AppRuntime>,
) -> Result<bool, String> {
    runtime.rename_tag(&old_name, &new_name).map_err(error_text)
}

#[tauri::command]
pub(crate) fn get_recent_tags(runtime: State<'_, AppRuntime>) -> Result<String, String> {
    runtime.get_recent_tags().map_err(error_text)
}

#[tauri::command]
pub(crate) fn set_recent_tags(json: String, runtime: State<'_, AppRuntime>) -> Result<(), String> {
    runtime.set_recent_tags(&json).map_err(error_text)
}

#[tauri::command]
pub(crate) fn list_selection_tags(
    selection: RowSelection,
    runtime: State<'_, AppRuntime>,
) -> Result<Vec<TagSelectionSummary>, String> {
    runtime.list_selection_tags(&selection).map_err(error_text)
}

#[tauri::command]
pub(crate) fn add_tags_to_selection(
    selection: RowSelection,
    tags: Vec<String>,
    runtime: State<'_, AppRuntime>,
) -> Result<TagMutationResult, String> {
    runtime
        .add_tags_to_selection(&selection, &tags)
        .map_err(error_text)
}

#[tauri::command]
pub(crate) fn remove_tags_from_selection(
    selection: RowSelection,
    tags: Vec<String>,
    runtime: State<'_, AppRuntime>,
) -> Result<TagMutationResult, String> {
    runtime
        .remove_tags_from_selection(&selection, &tags)
        .map_err(error_text)
}

#[tauri::command]
pub(crate) fn set_tags_for_row(
    row_id: i64,
    tags: Vec<String>,
    runtime: State<'_, AppRuntime>,
) -> Result<TagMutationResult, String> {
    runtime.set_tags_for_row(row_id, &tags).map_err(error_text)
}
